samza-core/src/main/java/org/apache/samza/table/remote/RemoteTable.java

```diff
@@ -90,33 +101,46 @@
    * @param writeFn {@link TableWriteFunction} for read operations
    * @param readRateLimiter helper for read rate limiting
    * @param writeRateLimiter helper for write rate limiting
-   * @param tableExecutor executor for issuing async requests
+   * @param rateLimitingExecutor executor for executing rate limiting
+   * @param readRetryPolicy read retry policy
+   * @param writeRetryPolicy write retry policy
+   * @param retryExecutor executor for invoking retries
    * @param callbackExecutor executor for invoking async callbacks
    */
-  public RemoteTable(String tableId, TableReadFunction readFn, TableWriteFunction writeFn,
-      TableRateLimiter<K, V> readRateLimiter, TableRateLimiter<K, V> writeRateLimiter,
-      ExecutorService tableExecutor, ExecutorService callbackExecutor) {
+  public RemoteTable(
+      String tableId,
+      TableReadFunction readFn,
+      TableWriteFunction writeFn,
+      TableRateLimiter<K, V> readRateLimiter,
+      TableRateLimiter<K, V> writeRateLimiter,
+      ExecutorService rateLimitingExecutor,
+      TableRetryPolicy readRetryPolicy,
+      TableRetryPolicy writeRetryPolicy,
+      ScheduledExecutorService retryExecutor,
+      ExecutorService callbackExecutor) {
+
     super(tableId);
-    Preconditions.checkNotNull(readFn, "null read function");
+    Preconditions.checkNotNull(readFn, "null readFn");
+
     this.readFn = readFn;
     this.writeFn = writeFn;
     this.readRateLimiter = readRateLimiter;
     this.writeRateLimiter = writeRateLimiter;
-    this.tableExecutor = tableExecutor;
+    this.rateLimitingExecutor = rateLimitingExecutor;
```

Comment:
You get better encapsulation by moving the rateLimits and retries entirely out of the RemoteTable. 

Here's one organization that comes to mind:

BaseRemoteTable <-- a basic RWTable implementation using reader functions and writer functions 
RetriableTable <- wraps a table with retries
RateLimitedTable <- wraps a table with rate-limits

A factory class like `RemoteTableProvider` can then instantiate the desired `RemoteTable` instance by wrapping it with the right set of decorators.

Reply:
Please see reply regarding sync/asncy